python/ray/data/_internal/execution/interfaces/physical_operator.py

```diff
@@ -273,7 +275,7 @@ def num_outputs_total(self) -> int:
             return self._estimated_num_output_bundles
         if len(self.input_dependencies) == 1:
             return self.input_dependencies[0].num_outputs_total()
```

Comment:
* should we remove the above if statement? this is the code logic that "assumes that each read task produces exactly one block"
* also, I'm thinking if we should make this method abstract to force each subclass to have a reasonable implementation.

Reply:
> should we remove the above if statement? this is the code logic that "assumes that each read task produces exactly one block"
Yeah, let me remove the if block, thanks.

In terms of forcing each operator to implement the method, I think for the majority of operators would simply return `self._estimated_num_output_bundles` if available, otherwise `None`. As a followup PR, I can refactor the individual logic to calculate `self._estimated_num_output_bundles` for operators into the `num_outputs_total()` method (for example, move [this logic](https://github.com/ray-project/ray/pull/46601/files#diff-4dd8434d27e556214b1d86755a50c1d9b4229bbea168dcb53c0d9b97cee67ca4R86-R92) into `num_outputs_total()` method). Let me know if you'd rather I include it in this PR instead.

Reply:
Actually I included the necessary changes in this PR, turned out to be simpler than I originally thought. For the default behavior, I included in `PhysicalOperator.num_outputs_total()` which returns `self._estimated_num_output_bundles`. Operators like `AllToAllOperator`, `Limit`, `Union`, and `Zip` have their own implementation which adds some more logic.

Reply:
Looks much nicer. Can you update `OutputSplitter` as well?

Reply:
added logic to increment `self._estimated_num_output_bundles` in `OutputSplitter._get_next_inner()`. let me know if you have different logic in mind.